File: quota.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

// RateLimitInfo holds parsed Anthropic rate-limit response headers.
type RateLimitInfo struct {
	Provider          string
	RequestsLimit     int
	RequestsRemaining int
	RequestsReset     time.Time
	TokensLimit       int
	TokensRemaining   int
	TokensReset       time.Time
	InputTokensLimit     int
	InputTokensRemaining int
	InputTokensReset     time.Time
	CheckedAt         time.Time
	Model             string
}

// CheckAnthropicQuota makes a minimal 1-token API call and reads rate-limit headers.
// Uses the cheapest model (haiku) to minimise cost.
func CheckAnthropicQuota() (*RateLimitInfo, error) {
	apiKey, src := resolveAnthropicKey()
	if apiKey == "" {
		return nil, fmt.Errorf(
			"ANTHROPIC_API_KEY not found. Set it with: setx ANTHROPIC_API_KEY sk-ant-...",
		)
	}
	_ = src // used for debug if needed

	model := "claude-haiku-4-5-20251001"
	reqBody := fmt.Sprintf(
		`{"model":%q,"max_tokens":1,"messages":[{"role":"user","content":"0"}]}`,
		model,
	)

	req, err := http.NewRequest("POST",
		"https://api.anthropic.com/v1/messages",
		strings.NewReader(reqBody),
	)
	if err != nil {
		return nil, err
	}
	req.Header.Set("x-api-key", apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")
	req.Header.Set("content-type", "application/json")

	client := &http.Client{Timeout: 15 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("API request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 401 {
		return nil, fmt.Errorf(
			"401 Unauthorized\n" +
				"  Key: " + apiKey[:min(20, len(apiKey))] + "...\n" +
				"  The key may be expired or revoked.\n" +
				"  Get a new key at https://console.anthropic.com/settings/keys\n" +
				"  Then: setx ANTHROPIC_API_KEY sk-ant-...   (in cmd/powershell)",
		)
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("API error %s", resp.Status)
	}

	info := &RateLimitInfo{
		Provider:  "anthropic",
		Model:     model,
		CheckedAt: time.Now(),
	}

	h := resp.Header
	fmt.Sscanf(h.Get("anthropic-ratelimit-requests-limit"), "%d", &info.RequestsLimit)
	fmt.Sscanf(h.Get("anthropic-ratelimit-requests-remaining"), "%d", &info.RequestsRemaining)
	fmt.Sscanf(h.Get("anthropic-ratelimit-tokens-limit"), "%d", &info.TokensLimit)
	fmt.Sscanf(h.Get("anthropic-ratelimit-tokens-remaining"), "%d", &info.TokensRemaining)
	fmt.Sscanf(h.Get("anthropic-ratelimit-input-tokens-limit"), "%d", &info.InputTokensLimit)
	fmt.Sscanf(h.Get("anthropic-ratelimit-input-tokens-remaining"), "%d", &info.InputTokensRemaining)

	parse := func(s string) time.Time {
		t, _ := time.Parse(time.RFC3339, s)
		return t
	}
	info.RequestsReset = parse(h.Get("anthropic-ratelimit-requests-reset"))
	info.TokensReset = parse(h.Get("anthropic-ratelimit-tokens-reset"))
	info.InputTokensReset = parse(h.Get("anthropic-ratelimit-input-tokens-reset"))

	saveQuotaToDB(info)
	return info, nil
}

func saveQuotaToDB(info *RateLimitInfo) {
	if db == nil {
		return
	}
	pct := 100.0
	if info.TokensLimit > 0 {
		pct = float64(info.TokensRemaining) / float64(info.TokensLimit) * 100
	}
	_, _ = db.Exec(
		`INSERT OR REPLACE INTO quota (provider, model_id, remaining, checked_at) VALUES (?,?,?,?)`,
		info.Provider, "", pct, info.CheckedAt.Format(time.RFC3339),
	)
}

// resolveAnthropicKey tries env var only (crush.json fallback removed).
func resolveAnthropicKey() (key, source string) {
	if v := os.Getenv("ANTHROPIC_API_KEY"); v != "" {
		return v, "env:ANTHROPIC_API_KEY"
	}
	return "", ""
}

// FormatQuota renders the rate-limit info as a human-readable string.
func FormatQuota(info *RateLimitInfo) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Anthropic rate limits  (checked %s)\n\n",
		info.CheckedAt.Format("2006-01-02 15:04:05"))

	type row struct {
		label     string
		remaining int
		limit     int
		reset     time.Time
	}
	rows := []row{
		{"Requests ", info.RequestsRemaining, info.RequestsLimit, info.RequestsReset},
		{"Tokens   ", info.TokensRemaining, info.TokensLimit, info.TokensReset},
	}
	if info.InputTokensLimit > 0 {
		rows = append(rows, row{"Input tok", info.InputTokensRemaining, info.InputTokensLimit, info.InputTokensReset})
	}

	for _, r := range rows {
		if r.limit == 0 {
			continue
		}
		pct := float64(r.remaining) / float64(r.limit) * 100
		bar := quotaBar(pct, 28)

		resetIn := ""
		if !r.reset.IsZero() {
			d := time.Until(r.reset).Truncate(time.Second)
			if d < 0 {
				d = 0
			}
			resetIn = fmt.Sprintf("  reset in %s", d)
		}
		fmt.Fprintf(&b, "  %s  %8s / %-8s  %s  %5.1f%%%s\n",
			r.label,
			fmtInt(r.remaining), fmtInt(r.limit),
			bar, pct, resetIn,
		)
	}

	fmt.Fprintf(&b, "\n  Model used for check: %s\n", info.Model)
	return b.String()
}

func quotaBar(pct float64, width int) string {
	if pct < 0 {
		pct = 0
	}
	if pct > 100 {
		pct = 100
	}
	filled := int(pct / 100 * float64(width))
	color := "█"
	switch {
	case pct < 20:
		color = "▓" // visually different for low quota
	case pct < 50:
		color = "█"
	}
	return "[" + strings.Repeat(color, filled) + strings.Repeat("░", width-filled) + "]"
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// QuotaHistory reads the last saved quota from DB.
func QuotaHistory() string {
	if db == nil {
		return "DB unavailable\n"
	}
	rows, err := db.Query(
		`SELECT provider, model_id, remaining, checked_at FROM quota ORDER BY checked_at DESC LIMIT 10`,
	)
	if err != nil {
		return fmt.Sprintf("error: %v\n", err)
	}
	defer rows.Close()

	var b strings.Builder
	fmt.Fprintln(&b, "Cached quota (from DB):")
	fmt.Fprintf(&b, "%-12s  %-10s  %8s  %s\n", "Provider", "Model", "Remaining", "Checked at")
	fmt.Fprintln(&b, strings.Repeat("─", 60))
	found := false
	for rows.Next() {
		var prov, modelID, checkedAt string
		var remaining float64
		if err := rows.Scan(&prov, &modelID, &remaining, &checkedAt); err != nil {
			continue
		}
		if modelID == "" {
			modelID = "(all)"
		}
		fmt.Fprintf(&b, "%-12s  %-10s  %7.1f%%  %s\n", prov, modelID, remaining, checkedAt)
		found = true
	}
	if !found {
		fmt.Fprintln(&b, "  (no records yet — run: model-manager quota claude)")
	}
	return b.String()
}

func fmtInt(n int) string {
	s := fmt.Sprintf("%d", n)
	if len(s) <= 3 {
		return s
	}
	// insert commas
	out := make([]byte, 0, len(s)+len(s)/3)
	for i, c := range s {
		if i > 0 && (len(s)-i)%3 == 0 {
			out = append(out, ',')
		}
		out = append(out, byte(c))
	}
	return string(out)
}

File: quota.py
"""quota.py — Anthropic rate-limit checker"""
import os
import time
from datetime import datetime, timezone
from typing import Optional

import httpx


class RateLimitInfo:
    def __init__(self):
        self.provider = "anthropic"
        self.requests_limit = 0
        self.requests_remaining = 0
        self.requests_reset: Optional[datetime] = None
        self.tokens_limit = 0
        self.tokens_remaining = 0
        self.tokens_reset: Optional[datetime] = None
        self.input_tokens_limit = 0
        self.input_tokens_remaining = 0
        self.input_tokens_reset: Optional[datetime] = None
        self.checked_at = datetime.now(timezone.utc)
        self.model = "claude-haiku-4-5-20251001"


def check_anthropic_quota() -> RateLimitInfo:
    api_key = os.environ.get("ANTHROPIC_API_KEY", "")
    if not api_key:
        raise ValueError(
            "ANTHROPIC_API_KEY not found. Set with: setx ANTHROPIC_API_KEY sk-ant-..."
        )

    model = "claude-haiku-4-5-20251001"
    body = {
        "model": model,
        "max_tokens": 1,
        "messages": [{"role": "user", "content": "0"}],
    }

    with httpx.Client(timeout=15) as client:
        resp = client.post(
            "https://api.anthropic.com/v1/messages",
            json=body,
            headers={
                "x-api-key": api_key,
                "anthropic-version": "2023-06-01",
                "content-type": "application/json",
            },
        )

    if resp.status_code == 401:
        raise ValueError(
            "401 Unauthorized\n"
            f"  Key: {api_key[:20]}...\n"
            "  The key may be expired or revoked.\n"
            "  Get a new key at https://console.anthropic.com/settings/keys"
        )
    if resp.status_code != 200:
        raise ValueError(f"API error {resp.status_code} {resp.reason_phrase}")

    info = RateLimitInfo()
    h = resp.headers
    info.requests_limit = _parse_int(h.get("anthropic-ratelimit-requests-limit", "0"))
    info.requests_remaining = _parse_int(h.get("anthropic-ratelimit-requests-remaining", "0"))
    info.tokens_limit = _parse_int(h.get("anthropic-ratelimit-tokens-limit", "0"))
    info.tokens_remaining = _parse_int(h.get("anthropic-ratelimit-tokens-remaining", "0"))
    info.input_tokens_limit = _parse_int(h.get("anthropic-ratelimit-input-tokens-limit", "0"))
    info.input_tokens_remaining = _parse_int(h.get("anthropic-ratelimit-input-tokens-remaining", "0"))
    info.requests_reset = _parse_rfc3339(h.get("anthropic-ratelimit-requests-reset", ""))
    info.tokens_reset = _parse_rfc3339(h.get("anthropic-ratelimit-tokens-reset", ""))
    info.input_tokens_reset = _parse_rfc3339(h.get("anthropic-ratelimit-input-tokens-reset", ""))

    # Save to DB
    try:
        from db import get_db, ensure_model_key, set_quota
        conn = get_db()
        mk = ensure_model_key(conn, "anthropic", model)
        pct = (info.tokens_remaining / info.tokens_limit * 100) if info.tokens_limit > 0 else 100
        set_quota(conn, "anthropic", mk, pct)
        conn.close()
    except Exception:
        pass

    return info


def format_quota(info: RateLimitInfo) -> str:
    now = info.checked_at
    lines = [f"Anthropic rate limits  (checked {now.strftime('%Y-%m-%d %H:%M:%S')})", ""]

    rows = [
        ("Requests ", info.requests_remaining, info.requests_limit, info.requests_reset),
        ("Tokens   ", info.tokens_remaining, info.tokens_limit, info.tokens_reset),
    ]
    if info.input_tokens_limit > 0:
        rows.append(("Input tok", info.input_tokens_remaining, info.input_tokens_limit, info.input_tokens_reset))

    for label, remaining, limit, reset in rows:
        if limit == 0:
            continue
        pct = remaining / limit * 100 if limit else 0
        bar = _quota_bar(pct, 28)
        reset_str = ""
        if reset:
            d = (reset - datetime.now(timezone.utc)).total_seconds()
            if d < 0:
                d = 0
            h, m = int(d // 3600), int((d % 3600) // 60)
            reset_str = f"  reset in {h}h{m:02d}m"
        lines.append(f"  {label}  {remaining:>8,} / {limit:<8,}  {bar}  {pct:5.1f}%{reset_str}")

    lines.append(f"\n  Model used for check: {info.model}")
    return "\n".join(lines)


def quota_history(days: int = 7) -> str:
    try:
        from db import get_db
        conn = get_db()
        rows = conn.execute(
            "SELECT provider, model_key, remaining, checked_at FROM quota ORDER BY checked_at DESC LIMIT 10"
        ).fetchall()
        conn.close()
    except Exception:
        return "DB unavailable\n"

    if not rows:
        return "  (no records yet — run: mm_quota)\n"

    lines = ["Cached quota (from DB):",
             f"{'Provider':<12}  {'Model':<22}  {'Remaining':>8}  {'Checked at':<20}",
             "─" * 70]
    for r in rows:
        lines.append(f"{r['provider']:<12}  {(r['model_key'] or '(all)'):<22}  {r['remaining']:>7.1f}%  {r['checked_at'][:19]:<20}")
    return "\n".join(lines)


def _parse_int(s: str) -> int:
    try:
        return int(s)
    except (ValueError, TypeError):
        return 0


def _parse_rfc3339(s: str) -> Optional[datetime]:
    if not s:
        return None
    try:
        return datetime.fromisoformat(s.replace("Z", "+00:00"))
    except Exception:
        return None


_QUOTA_CHARS = ["▓", "█", "░"]


def _quota_bar(pct: float, width: int = 28) -> str:
    if pct < 0:
        pct = 0
    if pct > 100:
        pct = 100
    filled = int(pct / 100 * width)
    color = "▓" if pct < 20 else "█"
    return "[" + color * filled + "░" * (width - filled) + "]"
